### ISSUE TITLE
tighten up rules for invalid use of aria-label and aria-labelledby

### ISSUE THREAD

**Issue by scottaohara** [validator/validator#1334]
With ARIA 1.2 certain roles are prohibited from being named by authors.

[ARIA in HTML](https://github.com/w3c/html-aria/pull/402) has identified the HTML elements which share these roles, and in some cases, have no corresponding ARIA role but too should be prohibited from being named by authors.  The following list of elements should flag an author error if `aria-label` or `aria-labelledby` are specified on them UNLESS a role which allows for naming is also specified.

**Note:** many of these roles already flag a warning for use of `aria-label`, but the message implies that if someone feels this is in error, to respond as such.  These warnings, as well as other elements from the following list which are not presently flagged would need to flag an **error** for using these attributes without an explicit role which allows naming from authors.

Many of the following elements map to ([`generic`](https://www.w3.org/TR/wai-aria-1.2/#generic)) or no corresponding role - often which does not adequately support naming from author.  Or, to roles which also explicitly prohibit naming from authors.

- a no href 
- abbr
- address
- area no href 
- b 
- bdi
- bdo
- body 
- caption ([`caption`](https://www.w3.org/TR/wai-aria-1.2/#caption))
- cite
- code ([`code`](https://www.w3.org/TR/wai-aria-1.2/#code))
- data 
- del ([`deletion`](https://www.w3.org/TR/wai-aria-1.2/#deletion))
- div 
- em ([`emphasis`](https://www.w3.org/TR/wai-aria-1.2/#emphasis))
- figcaption ([`caption`](https://www.w3.org/TR/wai-aria-1.2/#caption))
- footer (if not scoped to body)
- header (if not scoped to body)
- hgroup 
- i 
- ins ([`insertion`](https://www.w3.org/TR/wai-aria-1.2/#insertion))
- kbd 
- label
- legend
- mark ([`mark`](https://w3c.github.io/aria/#mark))
- p ([`paragraph`](https://www.w3.org/TR/wai-aria-1.2/#paragraph))
- pre 
- q
- rp 
- rt
- s 
- samp 
- small 
- span 
- strong ([`strong`](https://www.w3.org/TR/wai-aria-1.2/#strong))
- sub ([`subscript`](https://www.w3.org/TR/wai-aria-1.2/#subscript))
- sup ([`superscript`](https://www.w3.org/TR/wai-aria-1.2/#superscript))
- time ([`time`](https://www.w3.org/TR/wai-aria-1.2/#time))
- u 
- var

**Comment by pkra** [comment]
👍 for https://github.com/validator/validator/issues/1930

**Comment by sideshowbarker** [comment]
Does these same rules now also apply to use of `aria-braillelabel` and `aria-brailleroledescription`?

**Comment by pkra** [comment]
> Does these same rules now also apply to use of `aria-braillelabel` and `aria-brailleroledescription`?

Though `aria-braillelabel` is ARIA 1.3: yes. It has the same restrictions on roles as aria-label, so aria-in-html's changes for ARIA 1.2 have the same effect for brailleLabel.

(I'm guessing you didn't mean `brailleroledescription` but something like `brailleLabelledBy` - which doesn't exist.)

**Comment by sideshowbarker** [comment]
> > Does these same rules now also apply to use of `aria-braillelabel` and `aria-brailleroledescription`?
> 
> Though `aria-braillelabel` is ARIA 1.3: yes. It has the same restrictions on roles as aria-label, so aria-in-html's changes for ARIA 1.2 have the same effect for brailleLabel.

OK, thanks

> (I'm guessing you didn't mean `brailleroledescription`

Right

> but something like `brailleLabelledBy` - which doesn't exist.)

Yeah, I was just confused 😄

**Comment by sideshowbarker** [comment]
Per https://w3c.github.io/html-aria/#el-rp, `rp` is a _“Naming prohibited”_ element, right?

Is there a reason `rp` isn’t included in the list in the issue description here, and not included in the tests at https://w3c.github.io/html-aria/tests/prohibit-names.html?

**Comment by sideshowbarker** [comment]
At https://w3c.github.io/html-aria/#el-address and https://w3c.github.io/html-aria/#el-hgroup and https://w3c.github.io/html-aria/#el-label, the ARIA in HTML spec doesn’t identify `address`, `hgroup`, and `label` as _“Naming prohibited”_ — yet, those elements are included in the list in the issue description here, and in the tests at https://w3c.github.io/html-aria/tests/prohibit-names.html.

Is there something I’m missing, or does the ARIA in HTML spec need to be updated to identify those as _“Naming prohibited”_? Or what?

**Comment by sideshowbarker** [comment]
> The following list of elements should flag an author error if `aria-label` or `aria-labelledby` are specified on them **UNLESS a role which allows for naming is also specified.**

Can you please clarify exactly what _“a role which allows for naming”_ means?

Does it just mean the role _cannot_ be `caption`, `code`, `deletion`, `emphasis`, `generic`, `insertion`, `paragraph`, `presentation`, `strong`, `subscript`, or `superscript`?

If that’s not what it means, please instead either explicitly enumerate here exactly which roles allow naming, or else point me to somewhere else normative where I can find a single list of exactly what those roles are.

**Comment by pkra** [comment]
I think the confusion stems from the fact that Scott filed this issue in 2022 and the spec has evolved since then. 

For example in 2023 there's https://github.com/w3c/html-aria/pull/463 adjusting hgroup, address, rp and in 2024 https://github.com/w3c/html-aria/pull/556 adjusts label.

I would recommend following the spec and seeking clarification on the spec tracker so the ARIA WG can help clarify.

**Comment by scottaohara** [comment]
yup, that's exactly it.  

here's the newer issue about hgroup/address - https://github.com/validator/validator/issues/1626

maybe rp was a miss, or was clarified sometime in the years gap and i forgot to file an issue?  thanks @pkra for helping out with clarifying this / yes the spec is at least all up to date for any gaps in the pr

**Comment by sideshowbarker** [comment]
I’m confused about the following test case:

```html
    <div class=test-case>
    <h3><code>footer</code> not scoped to <code>body</code></h3>
    <article>
      <footer aria-label="aria">fail</footer>
      <footer aria-labelledby=name>fail</footer>
    </article>
    </div>
```
…and the corresponding one for `header`.

Why is `fail` the expected result for that?

At https://w3c.github.io/html-aria/#el-footer the ARIA in HTML spec says:

>  If not a descendant of an article, aside, main, nav or section element, or an element with role=article, complementary, main, navigation or region then role=[contentinfo](https://w3c.github.io/html-aria/#index-aria-contentinfo)
> 
> Otherwise, role=generic
>
> [Naming Prohibited](https://w3c.github.io/html-aria/#dfn-naming-prohibited) if exposed as generic.

So, what I see is:

1. That `footer` is a descendant of an `article`.
2. Therefore it does not have the implicit role `generic` (but instead has the implicit role `contentinfo`).
3. Therefore it is not _“exposed as `generic`”_.
4. Therefore it is not _“Naming prohibited”_.

What am I misunderstanding?

**Comment by scottaohara** [comment]
all of the test cases are contained within a `main` element, so none of them are scoped to the `body`.

but, all of this will be moot, since within the last year there has been new roles created to replace the generic fallback roles for header and footer (sectionheader and sectionfooter) - so once this html aam PR lands (https://github.com/w3c/aria/pull/2543) then header and footer will not have a name prohibited condition anymore, as both of their implicit roles will accept naming.  

tldr; you probably want to ignore the conditions for header and footer since in the past 3.5 years things changed :D

**Comment by sideshowbarker** [comment]
@pkra @scottaohara I’ve landed a change for this and pushed it to https://validator.w3.org/nu/

Please take a look there at and the changes in https://github.com/validator/validator/commit/fb1bf7be70809c950595b6d7cd98b8aa51ed2656 and let me know what I missed.
